### ISSUE TITLE
deep_copy empty UnorderedMap segfaults

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#2625]
This code segfaults at the deep_copy() call:
```c++
#include <Kokkos_Core.hpp>
#include <Kokkos_UnorderedMap.hpp>

using ExecutionSpace = Kokkos::DefaultExecutionSpace;

using Key = int;
using Value = int;

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {

  using Map = Kokkos::UnorderedMap<Key,Value,ExecutionSpace>;

  Map m{};
  Map n{};
  /* n = Map{m.capacity()}; // this would segfault too*/
  n.rehash(m.capacity());
  Kokkos::deep_copy(n,m);

  } Kokkos::finalize();
}

```
However, it runs fine without the `n.rehash(m.capacity());` line. Both maps' `capacity()` return `0`.

**Comment by Char-Aznable** [comment]
This is compiled with clang 9 and runs on CPU.

**Comment by jeffmiles63** [comment]
Map m{}; contains an set of uninitialized internal objects (all default constructed).
n.rehash(); initializes the internal objects ( with un-intialized data).
deep_copy(n,m) creates a new map with initialized internal objects (with un-initialized data) and tries to copy the contents of the uninitialized m to the new n.

The real bug is that a default constructed View with static extents does not have any data allocated and data() is a nullptr, but a View with static extents that is constructed with a label has data and a valid pointer behind data().  The DeepCopy<>() is what segfaults because the internal m_scalars view is static and it tries to copy using the following ( <valid ptr>, nullptr, sizeof(int)*3 ) 

This too will segfault 
```C++
     using test_view = Kokkos::View<int[3]>;
     test_view av = test_view("av");
     test_view bv;
     Kokkos::Impl::DeepCopy<test_view::memory_space, 
                                                 test_view::memory_space>(
             av.data(),bv.data(), sizeof(int)*3);
```
 
Workaround Notes:
if  you rehash both n and m then it works because both maps are in the same state.

**Comment by jeffmiles63** [comment]
Follow up question on fix:  Should we change the behavior of `Kokkos::View<int[3]>()` or should we handle the case of `DeepCopy( [ValidPtr], nullptr, sizeof(int)*3 )`

**Comment by Char-Aznable** [comment]
It seems something like `Kokkos::View<int[3]>()` is not really a valid object. What is the use case of having default constructed view without label and therefore without ref counting? @crtrott That being said, I have a lot of app where the default constructor is using something like `Kokkos::View<int[3]>()`, which could hit a bug like this one.

**Comment by ndellingwood** [comment]
Looks like this was resolved by #2739
